### DISCUSSION TITLE
AIR Curiosity Questions

### DISCUSSION THREAD

**Discussion by Bilbosax** [airsdk/Adobe-Runtime-Support#1269] [General]
I absolutely love AIR.  It is hands down my favorite language to develop in.  But recently I have run into two things that make me consider trying other platforms - performance and native appearance.  Now, I am asking these questions here respectfully, I am not an advanced coder who understands the complexities of  a runtime and a compiler, so please, no offense is intended.

First, performance.  AIR packages everything with a runtime that seems to be the bottleneck in its performance.  My understanding is that React Native, Flutter, Xamarin, and Haxe all have the ability to fully compile down to native code.

Also, although I love to be able to skin components to give my app a branded look, NONE of my apps look like they belong on the device, none of them look "native".  When it comes to business apps, this is important to users.  Again, my understanding is the React Native, Xamarin, and Flutter all have the ability to build with native components like buttons, textinputs, datepickers, etc.

So my questions are:

1)  Why can't AIR compile to native code at this point in the game, and will it ever be able to?
2)  Why can't AIR create native components to give a more "native" appearance?  Is it a future possibility, or is it impossible for this solution to handle?

**Comment by dgurney75** [comment]
If you want your app to have native looking components, then you need a library of designed elements to start with. I work in Adobe Animate for this specific reason. I use the latest Adobe XD iOS and material design libraries to skin my functional components with. Since the Animate IDE is a visual interface that seemlessly plays nicely with other Adobe products, this works great. You just create 2 states for each component.

**Reply by Bilbosax** [reply]
@dgurney75 This is very interesting!  I had no idea that Adobe had native component designs, I kinda quit looking when they abandoned Flash thinking that they were totally out of that game.  Sadly, I do not have access to Animate.  You have to have Animate to be able to build components from these kits, don't you?

**Reply by dgurney75** [reply]
Here’s the link to the libraries I was talking about.

https://www.adobe.com/products/xd/features/ui-kits.html

**Comment by ajwfrost** [comment]
This is an interesting one. For the two points .. the native components I would agree with the above, there would need to be specially skinned components for this - I don't think those other frameworks you mention are actually using proper native components, I think they all have UI libraries that are just pretty closely mimicking them. But it's an interesting thought to see whether we could build up a display using some native UI mechanism rather than using the native AS3 components and displaylist mechanism.

For the "compile to native code" - well it compiles just-in-time from ActionScript Byte Code (within the SWF) into machine code. So once all the functions have gone through that process, then it's pretty close to being as fast as a natively compiled application. But there's no "whole program link optimization" step or similar that would have happened if you were compiling e.g. a C++ application, so it doesn't reach quite that speed... but it should be comparable to any of those other frameworks I would think. The challenge then is about the fact that there's the "just-in-time compilation" going on: the first time any AS3 function is entered, it's inspected, verified, converted to LIR, and turned into machine code before being run. Some functions (class initializers at least) are never JIT-compiled as the effort involved in this is considered excessive compared to the time spent just going through the interpreter.

On iOS, this process is instead done "ahead of time" so you do end up with a single compiled native binary that houses the application. Down-side: you can't dynamically load SWF files that contain ActionScript.

Folk had asked earlier about whether we could do this also for Android. Talking with Adobe they didn't think there would be that great a benefit, but it has to be said that the start-up time seems fairly significant to me on Android vs on iOS, so we're going to look into this again. And maybe the same could be done on desktop platforms too...

We did some work for a customer a few years ago who were looking to improve start-up time, to see whether we could 'save' the JIT-compiled output for loading faster the next time. But, the generated machine code has got some memory addresses hardcoded into it which would need to be updated every time it ran so that didn't go anywhere, we did a kind of 'hibernation' effect where the initial loading of the SWF file and creation of the memory regions was saved/restored which helped a bit but the other change we made was to strip out the 'housekeeping' i.e. all the process of loading in the XML descriptor file and working out what extensions to load, what profile it is, what the settings are etc. So this sort of thing is also an area we were wondering whether it could be improved.

Anyway: "Why can't AIR compile to native code" - it can I guess, since it does on iOS, and so we can look a little more into whether it can on other platforms. You would then of course need to have a linker, which is fine on Linux (gcc) and macOS (xcode) but perhaps more problematic on Window due to their tools licensing. Or we could push this sort of step into an online tool perhaps which would take away the complexities for developers... but it would likely only be something used for the "export release build" type approach. FYI we're looking at going the other way for debugging, i.e. where if someone edits their AS3 file and rebuilds, we can pick up the change in the SWF and update the code that's running so that it forgets whatever has been JIT-compiled and just reloads in the new ABC (so "edit and continue" or whatever you want to call it..)

Coming back to your comment on the "runtime": this, as well as containing the interpreter and JIT-compiler, houses the memory management module plus all the API implementations and engine to handle the display etc. So by its nature this is an abstraction over the top of the native mechanisms - otherwise it wouldn't be a platform-independent implementation. The alternative perhaps would be to use AS3 and have this "transpile" down to some native implementation, that is then compiled using the platform-specific tools for each operating system. But at that point you lose all the capabilities for the shapes/sprites/movieclips etc that have no native equivalent: for a "pure" AS3 app that maybe just has some bitmaps, then fine, but to have the power of the vector graphics and animation tools, I think the runtime is still needed...


Slightly rambling answer. But thanks for your comments and thoughts on what would be useful! It may be we need to go more for the "native compilation" and AOT approach anyway, particularly given where Apple are going with toolchains and security on macOS, and we have a task already set off to look into this further for the Android platform as a starting point.

cheers

**Reply by dgurney75** [reply]
If interested, I could create a base mobile app in Animate that has all of the components in it for iOS and Material Design. It could also include a basic login and Home Screen to get started with. It could also include all of supporting files to publish to to both platforms. Longer term, we could eleven provide templates for all of the native extensions that Distriqt offers. Anyway, let me know if this is something that would be useful.

**Reply by Bilbosax** [reply]
@dgurney75 I know personally I would be MORE than interested!  Having not worked with animate, I am not sure what options you would have for output.  I work with Flex, so MXML skins are ideal, but AS3 skins would work just as well.  Please let me know if outputting an iOS and Android library of component skins is something that you think would be possible!

**Reply by Bilbosax** [reply]
@ajwfrost Thank you for the detailed an educational response.  I am certain that you and your talented team know what are the concerns of your users and think daily of how to handle them.  I have just been talking to a lot of people about which other platforms I may consider for work to get a job, and in my city, mobile development is not a big draw so I have been considering C#.  This just brought up a lot of discussion about runtimes like CLR, AIR, Java, and others vs languages that compile down to pure native code. 

Anything that you can do to get your users to native speeds, with a native platform UI experience could do nothing but better your position in the cross-platform market.  Adobe seems to have UI kits that they have for Animate, but there are tons of us who don't use Animate.  Perhaps @dgurney75 can do something to help us out, but if not, it would be great if you could perhaps talk to Adobe and get them to offer something to their Flex and Starling users as well.  I don't think it would impact game design much, but for business apps, it would be indispensible!

**Reply by marchbold** [reply]
Would be interesting to get some UI packages together (potentially through apm?). 

I suppose a Feathers UI theme is an obvious first step for a lot of people and there are already some great ones of those around (@marpies had some great ones we've used in the past: https://feathers.marpies.com/themes/)

**Reply by marchbold** [reply]
Flex generally is a bit of a performance bottleneck (particularly on mobile) so if you are looking for speed I would suggest you move away from relying on that. Starling/Feathers + themes is a great alternative to flex and much much faster.

**Reply by Bilbosax** [reply]
@marchbold getting some UI packages together with a tool like apm I think sounds AWESOME!  I mean, the easier you make it for users to get the tools that they need, the more likely they will stick around IMO.

I write business apps.  I knew that Starling was faster for games due to the gpu and 3d capabilities, but for straight business apps, I was unaware that Feathers was a faster alternative than Flex.  Thanks for the heads-up.

**Reply by leossmith** [reply]
I just wanted to share my experience with working using the above mentioned technologies. Working both with AIR and Flutter, the AIR performance has never been even close to Flutter or the rest of the mentioned technologies, especially on Android. Adding on top of that the Flex framework and you get a very bad performance compared to other stacks.  I will post tomorrow some of the findings from Firebase performance, on the same app, using the same libraries. 

Not only that, but apk size is at least x4 with AIR which I am not sure if it is due to not compiling to 100% native code or something else. 

For example our Pokemon app compiled with AIR:
![image](https://user-images.githubusercontent.com/4300876/137588673-38516769-9044-4981-9828-eb609aef404d.png)

VS the same exact app after migrated to Flutter:
![image](https://user-images.githubusercontent.com/4300876/137588711-8e0f0429-a3ad-46d2-a20c-df3db1dc527b.png)

Both apps are x64 .apk

And if someone wants to try performance on their devices: [Huawei APK (AIR)](https://appgallery.huawei.com/app/C100574155), [Android (Flutter) ](https://play.google.com/store/apps/details?id=tech.simpleapps.pokecalculator)

**Reply by ajwfrost** [reply]
@leossmith I suspect the large size of that META-INF folder is caused by the way that native extensions were being packaged up into APKs. I don’t know why but it seems the whole ANE file is just extracted into this path which means often you end up with duplicate or irrelevant files in there. We’ve changed this now when we create the App Bundle files so I would hope it’s a lot better now.

But yes I would still expect an AIR app to be bigger because it includes the whole runtime library, including all the capabilities of AIR that you may not be using at all, and all the ActionScript API definitions and implementations.

So if we could switch to an AOT mechanism then it should improve both the startup time and mean that the linkers can optimise away a lot of unused code…

**Reply by leossmith** [reply]
@ajwfrost it is true that there is a significant difference in size for appbundle, but still not good enough. 

You can see here a comparison between the old size (AIR AAB) vs the new size (Flutter AAB):
![image](https://user-images.githubusercontent.com/4300876/137627996-56a4b313-c09c-4e71-9ea1-40911c8a5253.png)

Back to performance now, it looks like firebase performance reports a 366ms starting time for Flutter app and 1.85s for Flex app an average of 5 times slower. I will write some tests to compare basic usage times during the coming week but I doubt that the old app will be anywhere near the performance of the new app (also the reason we decided to move to a full migration). 

Point being, although I love AIR and helps me develop super fast, sadly it is slow and not comparable to modern stacks. I would probably compare it to cordova / ionic and not Flutter, react or xamarin.

**Reply by ajwfrost** [reply]
@leossmith thanks - so actually most of the size optimisations are not seen in the 'final' APK size, because the issues with the original APK mechanism are caused by duplication, and of course in a zip file, these are easily stripped down.

I suspect one of the main size issues is the AIR runtime itself - a large .so file, around 20MB or so. I'm not sure whether we would end up optimizing some of that out if we moved to AOT compilation, since it's always possible to use reflection in ActionScript hence we'd need to keep all AS3 API definitions within the library.

Start-up time, definitely should be improved by AOT compilation. I had a bit of a look at this over the weekend and some of it is easier than I thought, but building the library as a .so using AOT will be the challenging part...

**Reply by radubirsan** [reply]
> @marchbold getting some UI packages together with a tool like apm I think sounds AWESOME! I mean, the easier you make it for users to get the tools that they need, the more likely they will stick around IMO.
> 
> I write business apps. I knew that Starling was faster for games due to the gpu and 3d capabilities, but for straight business apps, I was unaware that Feathers was a faster alternative than Flex. Thanks for the heads-up.

Adobe Flex is a very powerful framework build for PCs not for mobile devices. I would never use it on mobile and expect to have small file size and native speed.

**Comment by denisgl7** [comment]
Performance on android is really not good. Everything works many times better on my ios with approximately the same characteristics of the phone

**Reply by ajwfrost** [reply]
We’d looked into a couple of specific performance issues that people had raised in the past on Android, and found that part of the problem is because AIR is a C++ library but it has to call back into Java to access a lot of the Android platform capabilities. This JNI mechanism is sadly a bit expensive and takes a little bit of time which can then start to add up..

But we can look again into the idea of ahead-of-time compilation so that there’s no ActionScript running at runtime, it’s all pre-compiled to machine code and linked directly with the runtime like the approach that was enforced by Apple’s rules…

**Reply by denisgl7** [reply]
After a long run of the application, it starts to slow down, although at first everything is perfectly smooth. Can I help you?
My google app is gaining popularity even without ads, but users don't like lags
https://play.google.com/store/apps/details?id=air.ru.sever.ElectricalEngineering&hl=ru&gl=US

**Reply by ajwfrost** [reply]
@denisgl7 are you able to capture a performance log using Adobe Scout? This may help to understand what's causing the slowdown. There shouldn't be anything inherent in the platform that causes a problem (at least, until you've been running for 24.8 days..) so I'm wondering whether it's just got to a point where the garbage collection is ending up running more frequently..

thanks

**Reply by denisgl7** [reply]
Can this be done on the android platform?

**Reply by ajwfrost** [reply]
Yes .. for slightly older Android versions, there is an "Adobe Scout" utility that you can download from the Play Store that you can then use to configure the IP address of the host machine. So, run Scout on a PC that's on the same wifi network as your phone, and enter its IP address into the Scout companion app. It should connect, and then any app that runs (if built with apk-debug or aab-debug) should connect.

On newer Android versions where the file system access rights have changed, you would need to package up the telemetry.cfg configuration file with the IP address in it, within with your apk/aab package (this file should be generated when you launch Scout, in your home folder).
See http://renaun.com/blog/2013/01/using-adobe-scout-on-mobile-without-the-scout-companion-app/

**Comment by Bilbosax** [comment]
@leossmith Thanks for running the size and speed tests, these confirm my findings as well. Pacakage sizes are bigger, and speed is slower.

@ajwfrost Again, I love AIR. It is a mature platform with the partnership of a great ANE developer in distriqt and Feathers/Starling to take care of most of our needs. I think the only 2 things that are lagging are native look, and native SPEED. I think if you take care of these two things, you would see the community grow, not shrink. If there were ANY WAY AT ALL, getting AOT compiled code would improve performance and shrink file sizes dramatically, and keep you competitive.

And @marchbold , if there were a way to use apm to install and update UI libraries so that we could just use themes to move between Android and Apple devices, that would make our apps “look” native and would be HUGE in keeping AIR relevant to developers

Just my thoughts. Success to everyone!

**Reply by marchbold** [reply]
Yeah it just poses the question what is a "theme"? :)  

Makes sense when you are using a UI library like feathers but AIR doesn't really have UI components. There are a series of libs available though (not sure many apart from feathers are maintained though), but obviously theming all of them would be a massively time consuming process. 

From memory these were two of the popular ones:

- Minimal Comps
- Mad Components

But perhaps just starting with some feathers themes may be a better idea?

**Reply by ajwfrost** [reply]
It has to be said, the idea of using Flex on mobile makes me shudder a little .. it's a lot of overhead, I can see why people would have wanted to develop business apps in it particularly when the Flash Builder Design View still existed, because it means you can get a decent looking UI up quickly. But to make things easy at the time of development, it generally means it has more overhead at runtime.. and Flex took that approach to extremes!

I'm wondering though whether the Apache Royale frameworks might be of use? This is basically like Flex (still uses MXML/AS3) but with different frameworks and component sets, written in a much more lightweight manner - essentially, the "lessons learnt, let's do it again but better" approach. I haven't kept massively up to date with what they'd doing but originally you could create apps using MXML and AS3 which then compiled potentially to both HTML5/JS and to SWF formats.

**Reply by davecates** [reply]
This is where we have used Feathers before. https://feathersui.com/learn/as3-starling/

Shame Josh isn't maintaining it anymore as it was awesome. Not quite 100% native components but very usable.

We also created our own internal MVC UI framework for apps which is still going strong and used in production apps.

This is one of the weaknesses of the Air platform - gives you awesome capabilities but no quick and easy way to get a standard UI and framework up and running in minutes.

**Reply by marchbold** [reply]
@davecates Why do you say Josh isn't maintaining it anymore? He's about to push feathers as a package in apm.

**Reply by Bilbosax** [reply]
@marchbold “Themes” was probably a poor choice of words on my part, as that would imply baking it into a framework. I guess what I meant was that it would be nice if some like @dgurney75 would export Animate’s UI kits for Android and iOS as he offered above, and then you guys could use apm to install these files for use on our systems to make our apps look more native as if we were using a theme. Just an easy way of installing the files quickly from a repository.

**Reply by marchbold** [reply]
@Bilbosax Ah right, so more like an image/artwork library. Yeah sure, could just be a swc with a heap of vector or bitmap artwork. Perfect project for Animate and then packaging / distributing with apm is easy.

**Reply by dgurney75** [reply]
Let me know what you need and I’ll see what I can do. I also could create a base application in Animate that has demos of the ANEs that uses different UI components as examples.

**Reply by Bilbosax** [reply]
@dgurney75 I have never used Animate, so I don't know exactly what it is capable of.  Also, a lot of people do not use Animate but instead use Flex, or Starling, or Feathers, so an Animate project would not help them entirely.  To make it useful to everybody, what I think sounds perfect is if you were able to use Animate to quickly export a bunch of Actionscript Skins for each component.  Place all the Anroid Actionscript skins in a folder, place all of the iOS Actionscript skins in a folder, and export all the image assets needed for the skins in a folder.  Then you could just hand them over to @marchbold  to add to his repository. That would allow us to make changes to the skins if we wanted to customize their color, etc.  Would it be possible to use Animate to do something like this?  Does this sound like a feasible solution to everybody else??

**Reply by marchbold** [reply]
I'd suggest using Animate to produce an fla that exports a swc which contains a range of visual assets as exported classes, this is fairly standard process and one we often use for our app skins. 

@dgurney75 Focus solely on visual assets and not making an app. That way people can decide how to use the assets. If you open source it in a github repository I can add in the apm package components for you.

**Reply by Bilbosax** [reply]
@dgurney75, listen to @marchbold . Exporting a swc is a much better solution.

**Reply by leossmith** [reply]
There is also [this Material for Flex](https://github.com/quick6black/flex-Android-Material-Skins) which Aaron tried to donate to apache Flex and I have used extensively. Maybe it will help someone or someone will be interested in making it in .swc for apm.

**Reply by 2jfw** [reply]
@marchbold: Just a thought - what do you think about adding some kind of new section to **airsdk.dev** where (useful) things like that (referring to a possible Components SWC) get listed? Or is this too far away from the idea behind that page?

**Reply by marchbold** [reply]
@2jfw Just a little cautious about having to maintain something like that, probably suggest we start something like an "awesome actionscript" repo: eg could fork and update this one: https://github.com/robinrodricks/awesome-actionscript3 ?

**Reply by 2jfw** [reply]
@marchbold Ah, yeah, this one repo, I remember. Anyways alright, I got your point!

**Reply by marchbold** [reply]
@2jfw Not that I am opposed to adding something to the airsdk.dev site, however I sort of am hoping the apm repository (https://repository.airsdk.dev) will become more useful for this as we get libs available there. Hoping that people will see this as the best way to distribute and integrate libs in to AIR apps.

**Reply by 2jfw** [reply]
@marchbold I see - wasn't aware of this! Sounds good and this looks like the proper place for such assets etc
